nixos/modules/services/x11/display-managers/gdm.nix

```diff
@@ -310,8 +310,11 @@ in
 
       gdm-autologin.text = ''
         auth      requisite     pam_nologin.so
-
         auth      required      pam_succeed_if.so uid >= 1000 quiet
+        ${lib.optionalString pamCfg.login.enableGnomeKeyring ''
+          auth       [success=ok default=1]      ${pkgs.gnome.gdm}/lib/security/pam_gdm.so
+          auth       optional                    ${pkgs.gnome-keyring}/lib/security/pam_gnome_keyring.so
+        ''}
```

Comment:
@dawidd6 Did anyone from the NixOS's Security Team reviewed this? Referencing this in the chat kinda caused [some people to freak out seeing that](https://matrix.to/#/!RRerllqmbATpmbJgCn:nixos.org/$zIlhGMcyYcTEBZb-8HKVANqYalQFjJ80G9cgguwluA4?via=nixos.org&via=matrix.org&via=tchncs.de) and parsing secrets from privileged LUKS to the desktop for non-root users to have in their gnome-keyrings is kinda making me very nervous..

CC NixOS Security Team: @mweinelt @risicle @LeSuisse

Reply:
I don't know. PR was open for half a year, anyone could review it. AFAIK similar configuration can be found in other distros.

Reply:
@Kreyren The LUKS password is not being added to their gnome-keyring. A PAM module is reading the LUKS password from the kernel keyring and using it to unlock the gnome-keyring if possible. AFAIU, the user is never given access to the LUKS password.

Reply:
I only checked that PAM configuration itself should not be insecure, assuming the modules themself are secure. I also skimmed the `pam_gdm.so` module but not the `pam_gnome_keyring.so` (putting my hope in “enough eyeballs” of other distros).

If I understand the [`pam_gdm.so` code](https://gitlab.gnome.org/GNOME/gdm/-/blob/46.2/pam_gdm/pam_gdm.c) correctly, it does not send the LUKS secret, just the password used to unlock it.

But looking at `pam_gnome_keyring.so`, it is not clear to me if the password cannot be exfiltrated.


1. `pam_sm_authenticate` will try to call `unlock_keyring` https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L889
2. which determines the control file https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L680
    - This will be e.g. `/run/user/1000/keyring/control` socket owned by the user. Could some rogue service take it over?
    - Currently, I believe, we start the services with XDG autostart. Is that actually the case? Can that be hijacked?
3. Then it will call `gkr_pam_client_run_operation`: https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L690
4. which will call https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-client.c#L417
5. which will try to connect to the daemon https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-client.c#L318
6. And then send the arguments to it <https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-client.c#L342>, which [include the password](https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L688).
7. If daemon is not present and we do not pass `auto_start` flag (which we [do not do for `auth` module type](https://github.com/NixOS/nixpkgs/blob/d871c735fe0f9e43e2c2874c801e7db72363ad5f/nixos/modules/services/x11/display-managers/gdm.nix#L316)), it stashes the password for session: https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L895 https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L841
    - Presumably this refers to `session` PAM module type.
8. It would try to repeat the whole unlock/start circle again https://gitlab.gnome.org/GNOME/gnome-keyring/-/blob/46.2/pam/gkr-pam-module.c#L943

By the way, we are going to replace `pam_gdm.so` with `pam_systemd_loadkeys.so`: https://github.com/NixOS/nixpkgs/pull/286587 But that should have no effect on this.